### ISSUE TITLE
Why choose model.embeddings over model.embeddings.word_embeddings when calculating LayerIntegratedGradients for Bert like models.

### ISSUE THREAD

**Issue by cdpierse** [pytorch/captum#624]
Hi, 

I've been playing with the idea of only using a models' `embeddings.word_embeddings` when calculating LIG for Bert like models rather than `embeddings`. However I have noticed that with some models the attribution scores can change quite a bit for certain words. 

I know that for a model like Bert the embeddings are often broken down into `word_embeddings`, `token_type_embeddings`, and `position_embeddings`. However, the only embedding layer that returns meaningful results attribution results is `word_embeddings`, if I use the token or position embeddings I get `nan` for all word attributions. 

What I was wondering is if there is any reason for models like bert, distilbert, etc to calculate LIG attributions along all the embedding rather than just word ?

Let me know if you need any clarification on what I'm trying to ask, I hope it's not too vague. Thanks !

**Comment by NarineK** [comment]
@cdpierse, we have BERT tutorial that looks into all three types of embeddings (word, token and position).
https://captum.ai/tutorials/Bert_SQUAD_Interpret (this tutorial was built for an older version of Bert. We need to make some adjustments there.)
The attributions w.r.t. other embeddings aren't nan-s. There must be something else wrong. Since word embedding has the strongest predictive power you'll find them more meaningful. But as you can see from our tutorials other embeddings have meaningful attribution scores too.

**Comment by cdpierse** [comment]
@NarineK  thanks for the response. I had see that example before but I think I've put together why I was getting such weird responses, I wasn't testing very thoroughly apparently. I've been using the distilbert model as an example model for lot of these tasks, primarily because it's smaller faster to test on etc. The problem is that `distilbert` doesn't accept `position_ids` or `token_ids` as inputs for its forward pass unlike most other model configs in the [HF library](https://huggingface.co/transformers/model_doc/distilbert.html#distilbertconfig).  Thanks again for the response!

**Comment by NarineK** [comment]
@cdpierse, thank you for the explanation. I haven't used distilbert before. If `position_ids` and `token_ids` aren't used in the forward pass then we can't have  attributions for them.

**Comment by cdpierse** [comment]
No problem, my solution around this was to create a very simple helper function that uses python's inspect package to check the signature of the model's forward function. I've included a version of it below in case it's needed at some point.

```python
import inspect

def check_forward_signature_accepts_parameter(model, parameter) -> bool:
    signature = inspect.signature(model.forward)
    parameters = signature.parameters
    return parameter in parameters

```

**Comment by cdpierse** [comment]
So I had another go at this making sure to use a bert model that accepts position and token_type ids in its forward and still can't seem to get attributions w.r.t to the position and token_type embeddings that aren't all 0 tensors or nan when we attempt to normalize the summary. 

I did a sanity check using the example you provided in #311  which can be found here https://colab.research.google.com/drive/1Lw3JTZio03VwPvSVFzLJmZ52oBRpo9ZM. 

Changing the line:
`lig = LayerIntegratedGradients(custom_forward, model.bert.embeddings)`  --> 
`lig = LayerIntegratedGradients(custom_forward, model.bert.embeddings.position_embeddings)` 
will produce the nan response I mentioned previously. This is because the summarize attributions function attempts to divide a all 0 attribution tensor by its norm. I haven't been able to find a way to get layer integrated gradients to give correct attributions for the position embeddings.  

I also noticed some other things though, when  using `model.bert.embeddings`as the embeddings to pass to LIG you have to set the `n_steps` in `lig.attribute` to 7000 to get to a meaningful convergence for the attributions. 

If you directly use `model.bert.embeddings.word_embeddings` or `model.get_input_embeddings()` which are equivalent as the embeddings for the algorithm you don't need to change `n_steps` from the default (50) to get the same meaningful results you get when running for 7000 steps with all embeddings. 

My guess is here is that since `position_embeddings` and `token_type_embeddings` return all 0 attributions many more steps need to be carried out before convergence is reached as they are effectively noise. The number of steps required had come up as a point of conversation in #311 so maybe this helps shed some light on why the results were hard to make sense of.

**Comment by NarineK** [comment]
@cdpierse,  do you have the forward function of that model? I think what is happening is that since we are not setting any values for position ids it is using all zeros for position ids. IG ends up interpolating from baseline(all zeros) to actual inputs (all zeros) which we don't set and it is probably defaulted to 0. I think we need to find a way to pass actual values for position ids.

**Comment by cdpierse** [comment]
@NarineK so I made a couple of changes to your notebook and changed it so that the custom forward and predict functions also handle position_ids which are then passed to Bert's forward function using a keyword:

```python
def custom_forward(inputs, position_ids):
    preds = predict(inputs, position_ids)
    return torch.softmax(preds, dim = 1)[:, 0]
```

```python
def predict(inputs, position_ids):
    return model(inputs, position_ids=position_ids)[0]
```

With these I was hoping that LIG would be able to pick up that the actual position ids were present when calculating attributions w.r.t to the position_embeddings. I ran it that configuration in this notebook https://colab.research.google.com/drive/1Lw3JTZio03VwPvSVFzLJmZ52oBRpo9ZM?usp=sharing and unfortunately it still hasn't been able to pick up on the position ids.

**Comment by NarineK** [comment]
Thank you, @cdpierse! I cannot access the colab notebook.  I think we need to look closer what the second argument of model is taking. We can do that by looking into the forward function of the model. You can also use python's `inspect` package for that.

**Comment by cdpierse** [comment]
@NarineK  Oops, sorry about that. I've changed the permissions now so you should be able to look at it. I also added a quick section in the notebook that prints the models signature here's what it looks like:
```
(input_ids=None, attention_mask=None, token_type_ids=None, position_ids=None, head_mask=None, inputs_embeds=None, labels=None, output_attentions=None, output_hidden_states=None, return_dict=None)
```
In the notebook I'm passing position_ids as a `kwarg` so position as an argument shouldn't be too important. I know from looking at the Hugging Face internals that if the attention mask is not passed it simply uses a Tensor of 1s in the correct shape.

**Comment by NarineK** [comment]
@cdpiersem you'll need to pass position_ids as inputs to the attribute method
```
attributions_main, delta_main = lig.attribute(inputs=(input_ids, position_ids),
                                    baselines=(ref_input_ids,ref_position_ids),
                                    n_steps=70,
                                    internal_batch_size=3,
                                    #additional_forward_args=(position_ids),
                                    return_convergence_delta=True)
``` 
I tried this and it worked for me

**Comment by cdpierse** [comment]
@NarineK  Ah, this worked perfectly for me, I should have seen that inputs can take multiple values. Thanks for helping me with this, it's a huge help. Much appreciated ! Do you think the original SQUAD example or notebook would benefit from this addition ? I could put together a PR if you think so.

**Comment by NarineK** [comment]
@cdpierse, in the SQUAD tutorial we have `Multi-Embedding attribution` section that does the same with the IntegratedGradients but that code is using old Bert version and currently needs adjustments. I think we can change that section to use LayerIG for three layers.

```lig = LayerInttegratedGradients(squad_pos_forward_func, [bert.embeddings.word_embeddings, bert.embeddings.position_embeddings, bert.embeddings. token_type_embeddings])```

Feel free to make that change and the adjustments needed for the latest Bert model.

Let me know if you'd like to work on that PR.

Thank you,
Narine

**Comment by cdpierse** [comment]
@NarineK I'm happy to have a go at this, I think it should be okay to do. I can work on a revision of the notebook first using LayerIG exclusively, I'll put in a draft PR when I have something. Thanks, Charles.

**Comment by NarineK** [comment]
Thank you @cdpierse! Yes, we can use LayerIG instead of IG but it is good to keep LayerConductance. In terms of LayerConductance you can use the trick I used here:
https://captum.ai/tutorials/Bert_SQUAD_Interpret2

**Comment by NarineK** [comment]
This was addressed in: #672
